### DISCUSSION TITLE
Question - 802.1Qcr asynchronous traffic shaper implementation

### DISCUSSION THREAD

**Discussion by guohaorui** [CoRE-RG/CoRE4INET#31] [Q&A] [answered]
Hi,
I looked through the issue: Question - 802.1Qbu frame preemption implementation (https://github.com/CoRE-RG/CoRE4INET/issues/24) and you said INET 4.4.x will include many TSN-related features with many examples and documentation. (Qcr is also included). Recently i am working on relevant work on ATS (Asynchronous Traffic Shaper, proposed in Qcr), so i am wondering the progress of qcr implementation work (including UBS, paternoster), or maybe could you give me some suggestions on implementing UBS and paternoster?

**Comment by Musteblume** [comment, marked as answer]
Hi,
you can have a look at it in their repo (https://github.com/inet-framework/inet/tree/topic/tsn). You need the newest OMNeT++ 6 Preview (https://omnetpp.org/download/preview) to use the TSN-features of this INET branch.

**Comment by guohaorui** [comment]
Thanks very much! I saw a few relevant ATS related files in showcase dir in INET branch
